src/transformers/models/clvp/modeling_clvp.py

```diff
@@ -0,0 +1,1338 @@
+    """,
+    CLVP_START_DOCSTRING,
+)
+class CLVPSpeechModelWithProjection(CLVPPreTrainedModel):
+    config_class = CLVPSpeechConfig
+    main_input_name = "speech_ids"
+
+    _no_split_modules = None
+
+    def __init__(self, config: CLVPSpeechConfig):
+        super().__init__(config)
+
+        self.speech_model = CLVPTransformer(config)
+
+        self.speech_projection = nn.Linear(config.hidden_size, config.projection_dim, bias=False)
+
+        # Initialize weights and apply final processing
+        self.post_init()
+
+    def get_input_embeddings(self) -> nn.Module:
+        return self.speech_model.token_embedding
+
+    def set_input_embeddings(self, value):
+        self.speech_model.token_embedding = value
+
+    @add_start_docstrings_to_model_forward(CLVP_SPEECH_INPUTS_DOCSTRING)
+    @replace_return_docstrings(output_type=CLVPSpeechModelOutput, config_class=CLVPSpeechConfig)
+    def forward(
+        self,
+        speech_ids: Optional[torch.Tensor] = None,
+        attention_mask: Optional[torch.Tensor] = None,
+        use_causal_attention_mask: Optional[bool] = False,
+        output_attentions: Optional[bool] = None,
+        output_hidden_states: Optional[bool] = None,
+        return_dict: Optional[bool] = None,
+    ) -> Union[Tuple, CLVPSpeechModelOutput]:
+        r"""
+        Returns:
+
+        Examples:
+
+        ```python
+        >>> import torch
+        >>> from transformers import CLVPSpeechModelWithProjection
+
+        >>> model = CLVPSpeechModelWithProjection.from_pretrained("susnato/clvp_dev")
+
+        >>> # TODO : after FeatureExtractor is implemented we need to change it to something lik FE.__call__(...)
+        >>> inputs = {"speech_ids": torch.tensor([[5, 62, 1, 5, 9, 10]]).long()}
+
+        >>> outputs = model(**inputs)
+        >>> speech_embeds = outputs.speech_embeds
+        ```
+        """
+
+        return_dict = return_dict if return_dict is not None else self.config.use_return_dict
+
+        speech_outputs = self.speech_model(
+            input_ids=speech_ids,
+            use_causal_attention_mask=use_causal_attention_mask,
+            attention_mask=attention_mask,
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+            return_dict=return_dict,
+        )
+
+        pooled_output = speech_outputs[1]  # pooled_output
+
+        speech_embeds = self.speech_projection(pooled_output)
+
+        if not return_dict:
+            outputs = (speech_embeds, speech_outputs[0]) + speech_outputs[2:]
+            return tuple(output for output in outputs if output is not None)
+
+        return CLVPSpeechModelOutput(
+            speech_embeds=speech_embeds,
+            last_hidden_state=speech_outputs.last_hidden_state,
+            hidden_states=speech_outputs.hidden_states,
+            attentions=speech_outputs.attentions,
+        )
```

Comment:
The only purpose of the `Projection Models` I think is - 
- If we just want to get the projections of either `speech` or `text` it would be great to have a specific model since we don't need to load the other part, so it would save huge amount of memory. 
 
I am currently thinking of adding the `GPT2LMHead` model in `CLVPSpeechModelWithProjection` since it's specific to `speech` so as of current design we must keep the both of them.

Reply:
I have removed both the `text` and `speech` models and created a single `CLVPTransformerWithProjection` as you said.